src/Adapter/Product/CommandHandler/GenerateProductCombinationsHandler.php

```diff
@@ -0,0 +1,153 @@
+
+    /**
+     * {@inheritdoc}
+     *
+     * @todo: multistore
+     */
+    public function handle(GenerateProductCombinationsCommand $command): array
+    {
+        $product = $this->getProduct($command->getProductId());
+        $generatedCombinations = $this->combinationGenerator->generate($command->getGroupedAttributeIds());
+
+        SpecificPriceRule::disableAnyApplication();
+
+        $combinationIds = $this->addCombinations($product, $generatedCombinations);
+
+        Product::updateDefaultAttribute($product->id);
+        SpecificPriceRule::enableAnyApplication();
+        SpecificPriceRule::applyAllRules([$product->id]);
+
+        return $combinationIds;
+    }
+
+    /**
+     * @param Product $product
+     * @param Iterator $generatedCombinations
+     *
+     * @return CombinationId[]
+     *
+     * @throws CombinationException
+     */
+    private function addCombinations(Product $product, Iterator $generatedCombinations): array
+    {
+        $addedCombinationIds = [];
+        foreach ($generatedCombinations as $generatedCombination) {
+            $combinationExists = $product->productAttributeExists($generatedCombination, false, null, true);
+
+            if ($combinationExists) {
+                continue;
+            }
+
+            try {
+                $addedCombinationIds[] = $this->addCombination($product, $generatedCombination);
+            } catch (PrestaShopException $e) {
+                throw new CombinationException(
+                    sprintf(
+                        'Error occurred when trying to add combination for product #%d. Combination: %s',
+                        $product->id,
+                        var_export($generatedCombination, true)
+                    ),
+                    $e
+                );
+            }
+        }
+
+        return $addedCombinationIds;
+    }
+
+    /**
+     * @param Product $product
+     * @param int[] $generatedCombination
+     *
+     * @return CombinationId
+     */
+    private function addCombination(Product $product, array $generatedCombination): CombinationId
+    {
+        $newCombination = new Combination();
+        $newCombination->id_product = $product->id;
+        $newCombination->default_on = 0;
+        $product->setAvailableDate();
+
+        if (!$newCombination->add()) {
+            throw new CannotAddCombinationException(sprintf(
+                'Failed adding new combination for product #%d. Combination: %s',
+                $product->id,
+                var_export($generatedCombination, true)
+            ));
+        }
+        $combinationId = (int) $newCombination->id;
+        $this->saveProductAttributeAssociation($combinationId, $generatedCombination);
+
+        return new CombinationId($combinationId);
+    }
+
+    /**
+     * @param int $combinationId
+     * @param array $attributeIds
+     */
+    private function saveProductAttributeAssociation(int $combinationId, array $attributeIds): void
```

Comment:
To avoid having orphan combinations, `saveProductAttributeAssociation` should return a boolean and you should check that in `addCombination` and throw an exception if something bad happens (and delete the combination maybe ?)

Reply:
Guess you are right, or ill just add db transaction and roll back in case of error.

Reply:
@matks, now that I moved Db transactions & combinations wrapping in separate PR this place is unstable I guess.